### PR TITLE
expression: refactor hybrid type expressions

### PR DESCRIPTION
**Pull request by zz-jason**
For `Column` and `Constant`, we only need to consider hybrid type in `EvalInt` and `EvalString`.

For two simple plus(decimal, decimal) expressions: `SELECT L_QUANTITY+L_DISCOUNT, L_TAX+L_EXTENDEDPRICE FROM LINEITEM;` can be reduced from 12.81s to 10.72s

### CODE REVIEW

**Comment by ngaut** [comment]
Good job.

File: expression/builtin_cast.go [hunk 1]
```diff
@@ -394,12 +404,12 @@ func (c *castAsJSONFunctionClass) getFunction(ctx context.Context, args []Expres
 	}
 	bf := newBaseBuiltinFunc(ctx, args)
 	bf.tp = c.tp
-	if IsHybridType(args[0]) {
-		sig = &builtinCastJSONAsJSONSig{bf}
-		sig.setPbCode(tipb.ScalarFuncSig_CastJsonAsJson)
-		return sig, nil
+	var argTp types.EvalType
+	if args[0].GetType().Hybrid() {
+		argTp = types.ETInt
```
Comment (coocood) [inline expression/builtin_cast.go]: The argTp should be `ETString`?
```
create table e (a enum("a", "b", "c"));
mysql> insert e values ("a"), ("b"), ("c");
mysql> select cast(a as JSON) from e;
+-----------------+
| cast(a as JSON) |
+-----------------+
| "a"             |
| "b"             |
| "c"             |
+-----------------+
```

File: expression/builtin_cast.go [hunk 2]
```diff
@@ -348,12 +358,12 @@ func (c *castAsDurationFunctionClass) getFunction(ctx context.Context, args []Ex
 	}
 	bf := newBaseBuiltinFunc(ctx, args)
 	bf.tp = c.tp
-	if IsHybridType(args[0]) {
-		sig = &builtinCastDurationAsDurationSig{bf}
-		sig.setPbCode(tipb.ScalarFuncSig_CastDurationAsDuration)
-		return sig, nil
+	var argTp types.EvalType
+	if args[0].GetType().Hybrid() {
+		argTp = types.ETInt
```
Comment (coocood) [inline expression/builtin_cast.go]: cast an enum value to duration type is always zero.
So the ETInt type may be incorrect.

File: expression/builtin_cast.go [hunk 3]
```diff
@@ -302,12 +312,12 @@ func (c *castAsTimeFunctionClass) getFunction(ctx context.Context, args []Expres
 	}
 	bf := newBaseBuiltinFunc(ctx, args)
 	bf.tp = c.tp
-	if IsHybridType(args[0]) {
-		sig = &builtinCastTimeAsTimeSig{bf}
-		sig.setPbCode(tipb.ScalarFuncSig_CastTimeAsTime)
-		return sig, nil
+	var argTp types.EvalType
+	if args[0].GetType().Hybrid() {
+		argTp = types.ETInt
```
Comment (coocood) [inline expression/builtin_cast.go]: Maybe ETString is the correct type.

**Comment by coocood** [comment]
LGTM

**Pull request merged**